File: backend/src/entities/Installment.ts
import { Entity, PrimaryGeneratedColumn, Column, CreateDateColumn, UpdateDateColumn, ManyToOne, JoinColumn, Index } from 'typeorm';
import { Branch } from './Branch.js';
import { MachineSale } from './MachineSale.js';
import { Payment } from './Payment.js';

const decimalTransformer = {
  to: (value: number | null | undefined) => value,
  from: (value: string | number | null | undefined) => (value !== null && value !== undefined ? Number(value) : 0),
};

@Entity('installments')
@Index(['branchId', 'dueDate', 'isPaid'])
@Index(['saleId', 'installmentNo'])
export class Installment {
  @PrimaryGeneratedColumn('uuid')
  id!: string;

  @Column({ type: 'uuid' })
  branchId!: string;

  @ManyToOne(() => Branch, { onDelete: 'RESTRICT' })
  @JoinColumn({ name: 'branchId' })
  branch!: Branch;

  @Column({ type: 'uuid' })
  saleId!: string;

  @ManyToOne(() => MachineSale, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'saleId' })
  sale!: MachineSale;

  @Column({ type: 'uuid', nullable: true })
  paymentId?: string | null;

  @ManyToOne(() => Payment, { nullable: true, onDelete: 'SET NULL' })
  @JoinColumn({ name: 'paymentId' })
  payment?: Payment | null;

  @Column({ type: 'int' })
  installmentNo!: number;

  @Column({ type: 'timestamp' })
  dueDate!: Date;

  @Column({ type: 'decimal', precision: 12, scale: 2, transformer: decimalTransformer })
  amount!: number;

  @Column({ type: 'decimal', precision: 12, scale: 2, default: 0, transformer: decimalTransformer })
  paidAmount!: number;

  @Column({ type: 'boolean', default: false })
  isPaid!: boolean;

  @Column({ type: 'boolean', default: false })
  isWaived!: boolean;

  @Column({ type: 'varchar', length: 255, nullable: true })
  waiveReason?: string | null;

  @Column({ type: 'timestamp', nullable: true })
  paidDate?: Date | null;

  @Column({ type: 'varchar', length: 50, nullable: true })
  receiptNumber?: string | null;

  @CreateDateColumn()
  createdAt!: Date;

  @UpdateDateColumn()
  updatedAt!: Date;
}

File: backend-java/src/main/java/com/murabha/cloud/entity/Installment.java
package com.murabha.cloud.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.util.UUID;

@Entity
@Table(name = "installments", indexes = {
    @Index(name = "idx_installment_sale_id", columnList = "sale_id"),
    @Index(name = "idx_installment_due_date", columnList = "due_date"),
    @Index(name = "idx_installment_is_paid", columnList = "is_paid"),
    @Index(name = "idx_installment_payment_id", columnList = "payment_id"),
    @Index(name = "idx_installment_branch_id", columnList = "branch_id")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
public class Installment {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "sale_id", nullable = false)
    private UUID saleId;

    @JsonIgnoreProperties({"installments", "payments", "hibernateLazyInitializer", "handler"})
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "sale_id", insertable = false, updatable = false)
    private MachineSale sale;

    @Column(name = "payment_id")
    private UUID paymentId;

    @JsonIgnore
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "payment_id", insertable = false, updatable = false)
    private Payment payment;

    @Column(name = "installment_no", nullable = false)
    private Integer installmentNo;

    @Column(name = "due_date", nullable = false)
    private LocalDate dueDate;

    @Column(nullable = false, precision = 12, scale = 2)
    private BigDecimal amount;

    @Builder.Default
    @Column(name = "paid_amount", nullable = false, precision = 12, scale = 2)
    private BigDecimal paidAmount = BigDecimal.ZERO;

    @Builder.Default
    @Column(name = "is_paid", nullable = false)
    private Boolean isPaid = false;

    @Builder.Default
    @Column(name = "is_waived", nullable = false)
    private Boolean isWaived = false;

    @Column(name = "waive_reason")
    private String waiveReason;

    @Column(name = "paid_date")
    private LocalDate paidDate;

    @Column(name = "receipt_number", length = 64)
    private String receiptNumber;

    @Column(name = "payment_place", length = 64)
    private String paymentPlace;

    @Column(name = "branch_id")
    private UUID branchId;

    @JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "branch_id", insertable = false, updatable = false)
    private Branch branch;

    @Version
    @Column(name = "version")
    private Long version;

    @CreationTimestamp
    @Column(nullable = false, updatable = false)
    private Instant createdAt;

    @UpdateTimestamp
    @Column(nullable = false)
    private Instant updatedAt;
}